### PR TITLE
Draggable

### PR DESCRIPTION
**Pull request by Hixie**
Introduce a Draggable class that wraps all the logic of dragging
something and dropping it on a DragTarget.

Update examples/widgets/drag_and_drop.dart accordingly.

Make the performance/transition part of routes optional.

### CODE REVIEW

File: sky/packages/sky/lib/src/fn3/app.dart [hunk 1]
```diff
@@ -69,7 +69,7 @@ class AppState extends State<App> {
 
   Widget build(BuildContext context) {
     return new Theme(
-      data: config.theme,
+      data: config.theme ?? new ThemeData.fallback(),
```
Comment (abarth) [inline sky/packages/sky/lib/src/fn3/app.dart]: Yay

File: sky/packages/sky/lib/src/fn3/dialog.dart [hunk 2]
```diff
@@ -148,8 +148,9 @@ class DialogRoute extends Route {
     );
   }
 
-  void popState([dynamic result]) {
+  void popped([dynamic result]) {
```
Comment (abarth) [inline sky/packages/sky/lib/src/fn3/dialog.dart]: didPop?  We're not all that consistent about how we name these notification methods.

File: sky/packages/sky/lib/src/fn3/drag_target.dart [hunk 3]
```diff
@@ -103,21 +101,113 @@ class DragController {
...
+
+  Duration get transitionDuration => const Duration();
+  bool get isOpaque => false;
+  Widget build(Key key, NavigatorState navigator) {
+    return new Positioned(
+      left: _lastOffset.dx,
+      top: _lastOffset.dy,
+      child: new IgnorePointer(
+        child: new Opacity(
+          opacity: 0.5,
```
Comment (abarth) [inline sky/packages/sky/lib/src/fn3/drag_target.dart]: Show we have the opinion about the opacity or push that down into the feedback widget?

File: sky/packages/sky/lib/src/fn3/drag_target.dart [hunk 4]
```diff
@@ -103,21 +101,113 @@ class DragController {
...
+}
+
+class Draggable extends StatefulComponent {
+  Draggable({ Key key, this.navigator, this.data, this.child, this.feedback }): super(key: key) {
+    assert(navigator != null);
+  }
+  NavigatorState navigator;
+  dynamic data;
+  Widget child;
+  Widget feedback;
```
Comment (abarth) [inline sky/packages/sky/lib/src/fn3/drag_target.dart]: final

File: sky/packages/sky/lib/src/fn3/navigator.dart [hunk 5]
```diff
@@ -13,44 +13,58 @@ typedef Widget RouteBuilder(NavigatorState navigator, Route route);
...
+      if (_performance.status != desiredStatus)
+        _performance.play(direction);
+    }
   }
-  bool get isActuallyOpaque => _performance != null && _performance.isCompleted && isOpaque;
 
   bool get hasContent => true; // set to false if you have nothing useful to return from build()
+  bool get ephemeral => false; // set to true if you want to be automatically popped when another route is pushed or popped
+  bool get modal => true; // set to false if you don't mind lower layers being interactive while you're up (ephemeral must then be true also, since it's likely those layers will push and pop routes)
+  bool get isOpaque => false; // set to true if when the transition is complete, there's no reason to build and paint the route behind you
```
Comment (abarth) [inline sky/packages/sky/lib/src/fn3/navigator.dart]: I'd make these dartdoc comments

**Comment by abarth** [comment]
I can haz test?  LGTM w/ test

File: sky/packages/sky/lib/src/fn3/drag_target.dart (hunk 3 above)
Comment (Hixie) [inline sky/packages/sky/lib/src/fn3/drag_target.dart]: I plan to abandon the concept of a feedback widget very soon, replacing this with a use of mimic.

**Comment by Hixie** [comment]
Added test. I made a bunch of changes so would appreciate a quick second look.

File: sky/packages/sky/lib/src/fn3/drag_target.dart [hunk 6]
```diff
@@ -3,15 +3,85 @@
...
+
+class Draggable extends StatefulComponent {
+  Draggable({ Key key, this.navigator, this.data, this.child, this.feedback }): super(key: key) {
+    assert(navigator != null);
+  }
+  final NavigatorState navigator;
+  final dynamic data;
+  final Widget child;
+  final Widget feedback;
+  DraggableState createState() => new DraggableState();
```
Comment (abarth) [inline sky/packages/sky/lib/src/fn3/drag_target.dart]: In other places, we've been using a blank line after the constructor and before the createState.  Not sure if that makes much of a difference.

File: sky/packages/sky/lib/src/fn3/drag_target.dart [hunk 7]
```diff
@@ -3,15 +3,85 @@
...
+        _route = null;
+      }
+    );
+    _route.update(point);
+    config.navigator.push(_route);
+  }
+
+  void _updateDrag(sky.PointerEvent event) {
+    if (_route != null) {
+      config.navigator.setState(() {
```
Comment (abarth) [inline sky/packages/sky/lib/src/fn3/drag_target.dart]: This seems a little fiddly, but ok.

**Comment by abarth** [comment]
LGTM

Thanks for the dartdocs

File: sky/packages/sky/lib/src/fn3/drag_target.dart (hunk 6 above)
Comment (Hixie) [inline sky/packages/sky/lib/src/fn3/drag_target.dart]: done

File: sky/packages/sky/lib/src/fn3/drag_target.dart (hunk 7 above)
Comment (Hixie) [inline sky/packages/sky/lib/src/fn3/drag_target.dart]: Yeah, totally. I'm not really sure what to do about it. It's not the only place we do this.

**Pull request merged**